In [2]:
import numpy as np
import pandas as pd
from scipy.stats import norm, t
from IPython.display import display, Math, Latex

## 1
Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

In [4]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

#### ковариация

In [12]:
Mzk = np.mean(zp * ks)
Mz = np.mean(zp)
Mk = np.mean(ks)
print(f"Mzk = {Mzk}, Mz = {Mz}, Mk = {Mk}")
cov_zk = Mzk - Mz * Mk
print(f"cov(z, k) = {cov_zk:.3f}")

Mzk = 81141.7, Mz = 101.4, Mk = 709.9
cov(z, k) = 9157.840


In [14]:
np.cov(zp, ks, ddof=0)

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

#### коэффициент корреляции Пирсона

In [16]:
sz = np.std(zp, ddof=0)
sk = np.std(ks, ddof=0)
corr_zk = cov_zk / sz / sk
print(f"corr(z, k) = {corr_zk:.3f}")

corr(z, k) = 0.887


In [19]:
np.corrcoef(zp, ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [20]:
zp = pd.Series(zp)
ks = pd.Series(ks)

In [22]:
zp.corr(ks)

0.8874900920739163

In [23]:
ks.corr(zp)

0.8874900920739164

## 2
Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [25]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
n = len(iq)
a = 0.05

In [34]:
display(Math(r'\overline{X_n} - t_{\alpha/2,n} \dfrac{\sigma}{\sqrt{n}} \leq m \leq \overline{X_n} + t_{\alpha/2,n} \dfrac{\sigma}{\sqrt{n}}'))

<IPython.core.display.Math object>

In [39]:
x_n = np.mean(iq)
t_an = t.ppf(1 - a / 2, n-1)
sn = np.std(iq, ddof=1)
print(f"Xn = {x_n:.3f}\nt_an = {t_an:.3f}\ns = {sn:.3f}")

Xn = 118.100
t_an = 2.262
s = 10.546


In [40]:
[x_n - t_an * sn / np.sqrt(n), x_n + t_an * sn / np.sqrt(n)]

[110.55608365158724, 125.64391634841274]

Доверительный интервал **[110.556, 125.644]**

## 3
Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.

In [41]:
s = 5 # ср кв генеральной совокупности
n = 27
xn = 174.2
a = 0.05

In [42]:
display(Math(r'\overline{X_n} - Z \dfrac{\sigma}{\sqrt{n}} \leq m \leq \overline{X_n} + Z \dfrac{\sigma}{\sqrt{n}}'))

<IPython.core.display.Math object>

In [44]:
z = norm.ppf(1 - a / 2)
z

1.959963984540054

In [45]:
[xn - z * s / np.sqrt(n), xn + z * s / np.sqrt(n)]

[172.3140237765397, 176.08597622346028]

Доверительный интервал: **[172.314, 176.086]**